File: CSharp/ColoredThreads.cs
/*
	This program uses the Task pattern rather than threads
*/
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

class Program
{
	// Console lock
	public static object _consoleLock = new object();
	// Random
	public static Random _random = new Random();
	
	// Recursive thing!
	public static void RecursiveThing(int number, ConsoleColor color)
	{
		// Check number
		if (number < 0)
		{
			ColoredConsoleMesssage($"{color} thread complete!", color);
			return;
		}

		// Print to screen
		ColoredConsoleMesssage($"{color} thread, number is {number}.", color);

		// Wait random time
		Thread.Sleep(_random.Next(1, 10) * 1000);

		// Call RecursiveThing
		RecursiveThing(number - 1, color);
	}

	// Async wrapper for RecursiveThing()
	public static Task RecursiveThingAsync(int number, ConsoleColor color) =>
		Task.Run(() => RecursiveThing(number, color));

	// Lock the console and write the message with desired color
	public static void ColoredConsoleMesssage(string message, ConsoleColor color)
	{
		lock(_consoleLock)
		{
			Console.ForegroundColor = color;
			Console.WriteLine(message);
			Console.ForegroundColor = ConsoleColor.Gray;
		}
	}

	// Main thread
	public static void Main()
	{
		// Setup colors
		ConsoleColor[] colors =
		{
			ConsoleColor.Green,
			ConsoleColor.Blue,
			ConsoleColor.Yellow,
			ConsoleColor.Red,
			ConsoleColor.Cyan
		};

		// Create threads
		var threads =
			from color
			in colors
			select RecursiveThingAsync(10, color);

		// Kick off threads and wait for completion
		Task.WaitAll(threads.ToArray());

		// All done!
		Console.WriteLine("All threads complete!");
	}
}

File: FSharp/coloredThreads.fs
(*
    This program uses the F# async pattern rather than threads
*)
open System
open System.Threading

let consoleLock = Object()
let random = Random()

let colorName color =
    match color with
    | ConsoleColor.Green -> "Green"
    | ConsoleColor.Blue -> "Blue"
    | ConsoleColor.Yellow -> "Yellow"
    | ConsoleColor.Red -> "Red"
    | ConsoleColor.Cyan -> "Cyan"
    | _ -> "Unknown"

let coloredConsoleMessage color message =
    lock consoleLock (fun () ->
    Console.ForegroundColor <- color
    printf "%s\n" message
    Console.ForegroundColor <- ConsoleColor.Gray
    )

let rec recursiveThing number color =
    coloredConsoleMessage color (sprintf "%s thread, number is %d" (colorName color) number)
    Thread.Sleep(random.Next(1, 10) * 1000)

    match number with
    | 0 -> coloredConsoleMessage color (sprintf "%s thread complete!" (colorName color))
    | _ -> recursiveThing (number - 1) color

let recursiveThingAsync number color =
    async { recursiveThing number color }

[<EntryPoint>]
let main argv =
    let colors =
        [ConsoleColor.Green;
        ConsoleColor.Blue;
        ConsoleColor.Yellow;
        ConsoleColor.Red;
        ConsoleColor.Cyan]

    colors
        |> List.map (recursiveThingAsync 10)
        |> Async.Parallel
        |> Async.RunSynchronously
        |> ignore

    printf "All threads complete!\n"
    0 // return an integer exit code
